src/gui/mzroll/librarymanager.cpp

```diff
@@ -0,0 +1,333 @@
+            item->setText(1, dbName);
+            item->setText(2, numRecords);
+            entryExists = true;
+        }
+        ++iter;
+    }
+
+    if (!entryExists) {
+        QStringList rowItems = QStringList() << status
+                                             << dbName
+                                             << numRecords;
+        QTreeWidgetItem *item = new QTreeWidgetItem(rowItems);
+        item->setData(0, Qt::UserRole, QVariant::fromValue(database));
+        libraryTable->addTopLevelItem(item);
+    }
+}
+
+LibraryRecord LibraryManager::_recordForFile(const QString& filepath)
+{
+    auto dbName = mzUtils::cleanFilename(filepath.toStdString());
+    auto dbCompounds = DB.getCompoundsSubset(dbName);
+    LibraryRecord databaseRecord(filepath,
+                                 QString::fromStdString(dbName),
+                                 static_cast<int>(dbCompounds.size()));
+    return databaseRecord;
+}
+
+LibraryWorker* LibraryManager::_newWorker()
+{
+    LibraryWorker *worker = new LibraryWorker;
+    worker->moveToThread(&_workerThread);
+    connect(&_workerThread, &QThread::finished, worker, &QObject::deleteLater);
+    _workerThread.start();
+    return worker;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+LibraryRecord::LibraryRecord(const QJsonObject& json)
+{
+    absolutePath = "";
+    databaseName = "";
+    numberOfCompounds = 0;
+
+    if (json.contains("absolutePath") && json["absolutePath"].isString())
+        absolutePath = json["absolutePath"].toString();
+
+    if (json.contains("databaseName") && json["databaseName"].isString())
+        databaseName = json["databaseName"].toString();
+
+    if (json.contains("numberOfCompounds") && json["numberOfCompounds"].isDouble())
+        numberOfCompounds = json["numberOfCompounds"].toInt();
+}
+
+QJsonObject LibraryRecord::toJson() const
+{
+    QJsonObject json;
+    json["absolutePath"] = absolutePath;
+    json["databaseName"] = databaseName;
+    json["numberOfCompounds"] = numberOfCompounds;
+    return json;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+void LibraryWorker::saveRecord(const LibraryRecord& record)
+{
+    bool recordAlreadyExists = false;
+    auto entries = _loadEntries();
+    for (size_t i = 0; i < entries.size(); ++i) {
+        QJsonObject entry = entries[i].toObject();
+        LibraryRecord existingRecord(entry);
+        // record having the exact same file path should have the same stats
+        if (existingRecord.absolutePath == record.absolutePath) {
+            existingRecord.databaseName = record.databaseName;
+            existingRecord.numberOfCompounds = record.numberOfCompounds;
+            entries.replace(i, existingRecord.toJson());
+            recordAlreadyExists = true;
+        }
+    }
+
+    if (!recordAlreadyExists)
+        entries.append(record.toJson());
+
+    emit saveFinished(_saveEntries(entries));
```

Comment:
I don't see a connected slot for this. Why is this signal being emitted?

Reply:
Oh this is just for the sake of API completion and uniformity. In my head, if the worker emits a `loadFinished` signal, then it should also emit similar sibling signals. Maybe we'll find a use for them in the future.

Reply:
Got it. Makes sense to have this.